packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/bedrock/create-model-customization-job.ts

```
@@ -0,0 +1,542 @@
+    sfn.IntegrationPattern.REQUEST_RESPONSE,
+    sfn.IntegrationPattern.RUN_JOB,
+  ];
+
+  protected readonly taskMetrics?: sfn.TaskMetricsConfig;
+  protected readonly taskPolicies?: iam.PolicyStatement[];
+
+  private readonly integrationPattern: sfn.IntegrationPattern;
+  private _role: iam.IRole;
+
+  constructor(scope: Construct, id: string, private readonly props: BedrockCreateModelCustomizationJobProps) {
+    super(scope, id, props);
+    this.integrationPattern = props.integrationPattern ?? sfn.IntegrationPattern.REQUEST_RESPONSE;
+
+    this.validateStringLength('clientRequestToken', 1, 256, props.clientRequestToken);
+    this.validatePattern('clientRequestToken', /^[a-zA-Z0-9](-*[a-zA-Z0-9])*$/, props.clientRequestToken);
+    this.validateStringLength('customModelName', 1, 63, props.customModelName);
+    this.validatePattern('customModelName', /^([0-9a-zA-Z][_-]?)+$/, props.customModelName);
+    this.validateArrayLength('customModelTags', 0, 200, props.customModelTags);
+    this.validateStringLength('jobName', 1, 63, props.jobName);
+    this.validatePattern('jobName', /^[a-zA-Z0-9](-*[a-zA-Z0-9\+\-\.])*$/, props.jobName);
+    this.validateArrayLength('jobTags', 0, 200, props.jobTags);
+    this.validateArrayLength('validationData', 1, 10, props.validationData);
+    this.validateArrayLength('securityGroups', 1, 5, props.vpcConfig?.securityGroups);
+    this.validateArrayLength('subnets', 1, 16, props.vpcConfig?.subnets);
+
+    validatePatternSupported(this.integrationPattern, BedrockCreateModelCustomizationJob.SUPPORTED_INTEGRATION_PATTERNS);
+
+    if (!this.props.validationData && !this.props.hyperParameters?.['Evaluation percentage']) {
+      throw new ValidationError('validationData or Evaluation percentage hyperparameter must be provided.', this);
+    }
+
+    this._role = this.renderBedrockCreateModelCustomizationJobRole();
+    this.taskPolicies = this.renderPolicyStatements();
+
+    if (this.props.customModelKmsKey) {
```

It's important to note that addToResourcePolicy has no effect when used on a KMS key imported via fromKeyArn.

Reference: [aws-cdk-lib/aws-kms/key.ts#L163](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-kms/lib/key.ts#L163).

It's recommended to check the return value of the method to ensure the policy was actually applied.

---

Which is better to throw error or show warnings when `customModelKmsKey` is imported key? For now, I have implemented it to display warnings.

```ts
if (this.props.customModelKmsKey) {
      const poliyStatement = new iam.PolicyStatement({
        actions: ['kms:Decrypt', 'kms:GenerateDataKey', 'kms:DescribeKey', 'kms:CreateGrant'],
        resources: ['*'],
        principals: [new iam.ArnPrincipal(this._role.roleArn)],
        conditions: {
          StringEquals: {
            'kms:ViaService': [
              `bedrock.${Stack.of(this).region}.amazonaws.com`,
            ],
          },
        },
      });
      const result = this.props.customModelKmsKey.addToResourcePolicy(poliyStatement, true);

      // For the imported key, user must add the policy statement to the key policy
      if (result.statementAdded === false) {
        Annotations.of(this).addWarning(`You must update 'customModelKmsKey' resource policy to add the following statement:
        ${poliyStatement.toString()}`);
      }
    }
```

---

Thanks for updating the PR. I haven’t worked with this Bedrock job much before, but here’s my take: if missing permissions on the KMS key ultimately cause the job to fail, we should throw an exception to follow the fail-fast principle. On the other hand, if the lack of permission doesn’t block the job from running, then a warning message should suffice.

---

Sure. I've updated the implementation to throw an error.

---

LGTM, could you help me resolve the build errors of unused imports?